### PR TITLE
Remove duplicated ProgressBarIter

### LINKED ISSUE #138: support IndexedParallelIterator in Rayon
Right now Parallel progress bars are basically spinners unless one manually sets a len.
Rayon supports Indexed iterators with a .len() method that could be used automatically.

### LINKED ISSUE #151: mismatched types when returning ProgressBar.wrap_iter()
``` rust
use indicatif::*;

fn progressWith<It: Iterator>(steps: It, total: u64) -> ProgressBarIter<It> {

    let bar_style = indicatif::ProgressStyle::default_bar()
        .template("{spinner:.green} [{elapsed_precise}] [{wide_bar:40.cyan/blue}] {pos}/{len}  {percent}% ({eta})");

    let bar = ProgressBar::new(total)
        .with_style(bar_style);
    
    let res = bar.wrap_iter(steps.into_iter())
    // error[E0308]: mismatched types 
    res
    // ^^^^^ expected struct `indicatif::iter::ProgressBarIter`, found struct `indicatif::progress::ProgressBarIter`
}
```

Using this way:
``` rust
    let len = steps.len();
    let items = progressWith(steps, len);

    items.map(|row| { step_address(row) })
        .for_each(|(url)| {
            ...
        });

```

### LINKED ISSUE #242: Derive progress bar element count from ExactSizeIterator::len() and (rayon) IndexedParallelIterator
Hi mitsuhiko :-)

First of all, thanks for the very cool project! Two quick observations and potential improvement ideas (checking with you before going further, e.g., implementing/submitting a PR).

For progress bars on iterators, the total element count is currently determined by `size_hint` for std iterators (https://github.com/mitsuhiko/indicatif/blob/main/src/iter.rs#L11) and fixed to `0` for rayon parallel iterators (https://github.com/mitsuhiko/indicatif/blob/main/src/iter.rs#L98). This might not be optimal, because size_hint is brittle (could not be present), and having 0 by default for parallel iterators was a bit confusing for me and seems to have bitten others as well (https://github.com/mitsuhiko/indicatif/issues/195).

Improvement idea: implement `progress()` only for `ExactSizeIterator` and `IndexedParallelIterator` (where it can take the precise element count from the iterator), but not for normal `Iterator` and `ParallelIterator`? This would make the case where the element count is known from the iterator still easy, but force users users to supply an explicit element count else.

Does that make sense, any downsides to this? (Except for the breaking change, but I would argue this is intentional, if the iterator carries no size.)

### PR DESCRIPTION
**Pull request by mibac138**
Fixes #151, since bb26b3473ed970e12e815f1f965f107850ec7b0a also fixes #242 (cc @danleh), fixes #138

### CODE REVIEW

**Comment by mitsuhiko** [comment]
Still interested in this but unfortunately I cannot reopen the pull request because the main development branch is no longer called master.

**Comment by djc** [comment]
Would be happy to merge this if you rebase!

**Comment by mibac138** [comment]
@djc I have now resolved the merge conflicts and added some more improvements

**Comment by danleh** [comment]
@mibac138 Thanks a lot for the CC and the quick changes! I can comment only on #242, but there your changes are exactly what I had in mind in my question :-) Cool also, that you keep the old behavior around, behind `try_progress`.

File: src/iter.rs [hunk 1]
```diff
@@ -1,20 +1,24 @@
...
-    /// length using `Iterator::size_hint`.
-    fn progress(self) -> ProgressBarIter<Self> {
-        let n = match self.size_hint() {
-            (_, Some(n)) => n as u64,
-            _ => 0,
-        };
-        self.progress_count(n)
+    /// Wrap an iterator with default styling. Uses `Iterator::size_hint` to get length.
+    /// Returns `Some(..)` only if `size_hint.1` is `Some`. If you want to create a progress bar
+    /// even if `size_hint.1` returns `None` use `progress_count` or `progress_with` instead.
```
Comment (djc) [inline src/iter.rs:13]: Do you actually have a use case for this (or have you seen anyone else express a use case for it)? If not, I'd rather wait and see if someone complains about the new behavior. We can always add it later in a compatible release...

File: src/iter.rs [hunk 2]
```diff
@@ -25,205 +27,133 @@ where
...
 
-        next
+        item
     }
 }
 
-impl<S, T: Iterator<Item = S>> ProgressIterator for T {
-    fn progress_with(self, progress: ProgressBar) -> ProgressBarIter<Self> {
-        ProgressBarIter { it: self, progress }
+impl<T: ExactSizeIterator> ExactSizeIterator for ProgressBarIter<T> {
```
Comment (djc) [inline src/iter.rs:56]: Would you mind splitting this commit into a few separate ones (or separate PRs)? For example, one that moves rayon support to a separate file, one that moves `ProgressBarIter`, one that renames field names, and one with other changes? Keeping everything in the same commit makes it much harder for me to review.

**Review by djc — COMMENTED** [review]
These look like good improvements, but reviewing them as presented here is a little challenging. Would be happy to take another look if you can split stuff up a bit more.

File: src/iter.rs (hunk 2 above)
Comment (mibac138) [reply src/iter.rs:61]: Sure thing, it should be easier now.

**Comment by djc** [comment]
Awesome, thanks!

**Pull request merged**